File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/AbstractPriceFilter.java
package cz.fb.manaus.reactor.price;


import com.google.common.collect.Range;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.PriceComparator;
import cz.fb.manaus.core.model.Side;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class AbstractPriceFilter {

    private final int minCount;
    private final double bulldozeThreshold;
    private final Range<Double> priceRange;
    @Autowired
    private PriceBulldozer bulldozer;


    public AbstractPriceFilter(int minCount, double bulldozeThreshold,
                               Range<Double> priceRange) {
        this.minCount = minCount;
        this.bulldozeThreshold = bulldozeThreshold;
        this.priceRange = priceRange;
    }

    List<Price> getSignificantPrices(int minCount, List<Price> prices) {
        var bySide = prices.stream().filter(this::priceRangeFilter)
                .collect(Collectors.groupingBy(Price::getSide));
        var sortedBack = PriceComparator.ORDERING.immutableSortedCopy(
                bySide.getOrDefault(Side.BACK, List.of()));
        var sortedLay = PriceComparator.ORDERING.immutableSortedCopy(
                bySide.getOrDefault(Side.LAY, List.of()));
        var bulldozedBack = bulldozer.bulldoze(bulldozeThreshold, sortedBack);
        var bulldozedLay = bulldozer.bulldoze(bulldozeThreshold, sortedLay);
        var topBack = bulldozedBack.stream().limit(minCount);
        var topLay = bulldozedLay.stream().limit(minCount);
        return Stream.concat(topBack, topLay).collect(Collectors.toList());
    }


    private boolean priceRangeFilter(Price price) {
        return priceRange.contains(price.getPrice());
    }


    public List<Price> filter(List<Price> prices) {
        return getSignificantPrices(minCount, prices);
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/AbstractPriceFilter.kt
package cz.fb.manaus.reactor.price


import com.google.common.collect.Range
import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.PriceComparator
import cz.fb.manaus.core.model.Side
import org.springframework.beans.factory.annotation.Autowired

abstract class AbstractPriceFilter(private val minCount: Int,
                                   private val bulldozeThreshold: Double,
                                   private val priceRange: Range<Double>) {
    @Autowired
    private lateinit var bulldozer: PriceBulldozer

    internal fun getSignificantPrices(minCount: Int, prices: List<Price>): List<Price> {
        val bySide = prices.filter { this.priceRangeFilter(it) }.groupBy { it.side }
        val sortedBack = PriceComparator.ORDERING.immutableSortedCopy(
                bySide.getOrDefault(Side.BACK, emptyList()))
        val sortedLay = PriceComparator.ORDERING.immutableSortedCopy(
                bySide.getOrDefault(Side.LAY, emptyList()))
        val bulldozedBack = bulldozer.bulldoze(bulldozeThreshold, sortedBack)
        val bulldozedLay = bulldozer.bulldoze(bulldozeThreshold, sortedLay)
        val topBack = bulldozedBack.take(minCount)
        val topLay = bulldozedLay.take(minCount)
        return topBack + topLay
    }

    private fun priceRangeFilter(price: Price): Boolean {
        return priceRange.contains(price.price)
    }

    fun filter(prices: List<Price>): List<Price> {
        return getSignificantPrices(minCount, prices)
    }
}
